File: codeforces/292/292B.kt
package codeforces

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val In = BufferedReader(InputStreamReader(System.`in`))
    val graph = In.readLine()!!.split(" ").map {it.toInt()}
    val n = graph[0]
    val m = graph[1]
    val arr = List(n){0}.toMutableList()
    for( i in 0..m) {
        val edges = In.readLine()!!.split(" ").map {it.toInt()}
        arr[edges[0]] += 1
        arr[edges[1]] += 1
    }
    arr.sortedDescending()
    if(arr[0] > 2 && arr[1] == 1 && arr[n-1] == 1)
        println("star topology")
    else if(arr[0] == 2 && arr[n-1] == 1)
        println("bus topology")
    else if(arr[0] == 2 && arr[n-1] == 2)
        println("ring topology")
    else
        println("unknown topology")
}

File: codeforces/292/292B.py
B="bus"
R="ring"
S="star"
UKN="unknown"

adj=[]
nodes, edges = map(int, input().split(" "))

for i in range(nodes):
    adj.append([])
for i in range(edges):
    a,b= map(int,input().split(" "))
    adj[a-1].append(b)
    adj[b-1].append(a)
max=0
min= 99999999999999999
count =0
for i in (range(nodes)):
    count = len(adj[i])
    if count > max:
        max=count
    if count < min :
        min= count

if max == 2 and min == 1:
    print(B, "topology")

elif max == 2 and min == 2:
    print(R, "topology")
elif max == edges and min ==1:
        print(S, "topology")
else:
    print(UKN, "topology")